---
title: "C# : Syntaxe alternative pour les tests d'exception"
date: 2021-05-12 08:36:56
tags: C#
---

AJA qu'il existait une alternative pour tester les exceptions en C#. Depuis toujours j'utilisais la syntaxe suivante :

```csharp
[TestMethod]
[ExpectedException(typeof(DivideByZeroException))]
public void Division_ShouldThrowIfDividedByZero()
{
    var a = 1;
    var b = 0;
    var c = a / b;
}
```

Cette syntaxe est assez limit√©e car elle ne permet pas d'effectuer d'assertions sur l'exception lev√©e pendant l'ex√©cution du test.

Certes, on peut ins√©rer un bloc try/catch et effectuer des assertions comme ceci:

```csharp
[TestMethod]
[ExpectedException(typeof(DivideByZeroException))]
public void Division_ShouldThrowIfDividedByZero()
{
    try
    {
        var a = 1;
        var b = 0;
        var c = a / b;
    }
    catch(Exception e)
    {
        Assert.AreEqual("Attempted to divide by zero.", e.Message);
        throw; // throw n√©c√©ssaire pour que le d√©corateur Expected soit honor√©
    }
}
```

Mais ce n'est pas tr√®s √©l√©gant üò¢.

Or il existe une autre syntaxe bien plus pratique :

```csharp
[TestMethod]
public void Division_ShouldThrowIfDividedByZero()
{
    var ex = Assert.ThrowsException<DivideByZeroException>(() =>
    {
        var a = 1;
        var b = 0;
        var c = a / b;
    });
    Assert.AreEqual("Attempted to divide by zero.", ex.Message);
}
```

La lisibilit√© des tests est am√©lior√©e, les code reviews sont plus sereines et le monde se porte mieux. üåà
